pkg/operator/operator.go

```diff
@@ -0,0 +1,271 @@
+package operator
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/blang/semver"
+	"github.com/golang/glog"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	appsclientv1 "k8s.io/client-go/kubernetes/typed/apps/v1"
+	coreclientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	rbacclientv1 "k8s.io/client-go/kubernetes/typed/rbac/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	operatorsv1alpha1 "github.com/openshift/api/operator/v1alpha1"
+	operatorconfigclientv1alpha1 "github.com/openshift/cluster-kube-apiserver-operator/pkg/generated/clientset/versioned/typed/kubeapiserver/v1alpha1"
+	operatorconfiginformerv1alpha1 "github.com/openshift/cluster-kube-apiserver-operator/pkg/generated/informers/externalversions/kubeapiserver/v1alpha1"
+	"github.com/openshift/library-go/pkg/operator/v1alpha1helpers"
+	"github.com/openshift/library-go/pkg/operator/versioning"
+)
+
+const (
+	targetNamespaceName = "openshift-kube-apiserver"
+	workQueueKey        = "key"
+)
+
+type KubeApiserverOperator struct {
+	operatorConfigClient operatorconfigclientv1alpha1.KubeapiserverV1alpha1Interface
+
+	appsv1Client appsclientv1.AppsV1Interface
+	corev1Client coreclientv1.CoreV1Interface
+	rbacv1Client rbacclientv1.RbacV1Interface
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+}
+
+func NewKubeApiserverOperator(
+	operatorConfigInformer operatorconfiginformerv1alpha1.KubeApiserverOperatorConfigInformer,
+	namespacedKubeInformers informers.SharedInformerFactory,
+	operatorConfigClient operatorconfigclientv1alpha1.KubeapiserverV1alpha1Interface,
+	appsv1Client appsclientv1.AppsV1Interface,
+	corev1Client coreclientv1.CoreV1Interface,
+	rbacv1Client rbacclientv1.RbacV1Interface,
+) *KubeApiserverOperator {
+	c := &KubeApiserverOperator{
+		operatorConfigClient: operatorConfigClient,
+		appsv1Client:         appsv1Client,
+		corev1Client:         corev1Client,
+		rbacv1Client:         rbacv1Client,
```

Comment:
we probably don't need rbac here (appsv1 and corev1 seems ok for workloads/etc)...